crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -868,6 +839,136 @@ unsafe impl<'a, T: FromWorld + Send + 'static> SystemParam for Local<'a, T> {
     }
 }
 
+/// Types that can be used with [`Buffer<T>`] in systems.
+pub trait SystemBuffer: FromWorld + Send + 'static {
+    fn apply(&mut self, system_meta: &SystemMeta, world: &mut World);
+}
+
+/// A [`SystemParam`] that stores a buffer which gets applied to the [`World`] at the end of a stage.
+/// By using this to defer mutations, you can avoid mutable `World` data access within a system
+/// and allow it to run in parallel with more systems.
+///
+/// The most common example of this pattern is [`Commands`], which uses [`Buffer<T>`] to defer mutations.
+///
+/// [`Commands`]: crate::system::Commands
+///
+/// # Examples
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// // Tracks whether or not there is a threat the player should be aware of.
+/// #[derive(Resource, Default)]
+/// pub struct Alarm(bool);
+///
+/// // A threat from inside the settlement.
+/// #[derive(Component)]
+/// pub struct Criminal;
+///
+/// // A threat from outside the settlement.
+/// #[derive(Component)]
+/// pub struct Monster;
+///
+/// use bevy_ecs::system::{Buffer, SystemBuffer, SystemMeta};
+///
+/// // Uses deferred mutations to allow signalling the alarm from multiple systems in parallel.
+/// #[derive(Resource, Default)]
+/// struct AlarmFlag(bool);
+///
+/// impl SystemBuffer for AlarmFlag {
+///     fn apply(&mut self, system_meta: &SystemMeta, world: &mut World) {
+///         if self.0 {
+///             world.resource_mut::<Alarm>().0 = true;
+///             self.0 = false;
+///         }
+///     }
+/// }
+///
+/// // Sound the alarm if there is a criminal.
+/// fn alert_criminal(
+///     criminals: Query<&Criminal>,
+///     mut alarm: Buffer<AlarmFlag>
+/// ) {
+///     if criminals.iter().next().is_some() {
+///         alarm.0 = true;
+///     }
+/// }
+///
+/// // Sound the alarm if there is a monster.
+/// fn alert_monster(
+///     monsters: Query<&Monster>,
+///     mut alarm: Buffer<AlarmFlag>
+/// ) {
+///     if monsters.iter().next().is_some() {
+///         alarm.0 = true;
+///     }
+/// }
+///
+/// let mut world = World::new();
+/// world.init_resource::<Alarm>();
+///
+/// let mut stage = SystemStage::parallel();
+/// stage
+///     // These two systems have no conflicts and will run in parallel.
+///     .add_system(alert_criminal)
+///     .add_system(alert_monster);
+///
+/// // There are no criminals or monsters, so the alarm is not sounded.
+/// stage.run(&mut world);
+/// assert_eq!(world.resource::<Alarm>().0, false);
+///
+/// // Spawn a monster, which will cause the alarm to be sounded.
+/// world.spawn(Monster);
+/// stage.run(&mut world);
+/// assert_eq!(world.resource::<Alarm>().0, true);
+///
+/// // Reset the alarm.
+/// world.resource_mut::<Alarm>().0 = false;
+///
+/// // Spawn a criminal, which will cause the alarm to be sounded.
+/// world.spawn(Criminal);
+/// stage.run(&mut world);
+/// assert_eq!(world.resource::<Alarm>().0, true);
+/// ```
+pub struct Buffer<'a, T: SystemBuffer>(pub(crate) &'a mut T);
```

Comment:
One more suggestion: I think this is actually clearer if this is called `Deferred` instead. Buffers are used all over the place, and the Deferred system param will make this easier to parse at a glance.

Reply:
I don't disagree, but can you think of a good name for the `SystemBuffer` trait if we rename `Buffer`?

Reply:
`BufferParam` perhaps?

Reply:
Actually I think I prefer the current names. The type parameter refers to the data being stored, not the action being deferred. `Deferred<CommandQueue>` looks wrong to me because the queue isn't really being deferred, but `Buffer<CommandQueue>` looks right because we are using the queue as a buffer.

I can still change it if you want, though.

Reply:
Let's wait and see what others think. I do think it's clearer. If and only if we change to `Deferred` I would add it to the prelude.

Reply:
After trying this suggestion out, I agree that it's better.